Mathlib/Tactic/CC/Addition.lean

```diff
@@ -0,0 +1,2061 @@
+    return some aNeB₁
+  let b₁EqB ← getEqProof b b₁
+  match b₁EqB with
+  | none => return none -- failed to build proof
+  | some b₁EqB => mkAppM ``ne_of_ne_of_eq #[aNeB₁, b₁EqB]
+
+/-- If `e` is of the form `op e₁ e₂` where `op` is an associative and commutative binary operator,
+return the conanocal form of `op`. -/
+def isAC (e : Expr) : CCM (Option Expr) := do
+  let .app (.app op _) _ := e | return none
+  let ccs ← get
+  if let some cop := ccs.canOps.find? op then
+    let some b := ccs.opInfo.find? cop
+      | throwError "opInfo should contain all canonical operators in canOps"
+    return bif b then some cop else none
+  for (cop, b) in ccs.opInfo do
+    if ← pureIsDefEq op cop then
+      modify fun _ => { ccs with canOps := ccs.canOps.insert op cop }
+      return bif b then some cop else none
+  let b ←
+    try
+      let aop ← mkAppM ``Std.Associative #[op]
+      let some _ ← synthInstance? aop | failure
+      let cop ← mkAppM ``Std.Commutative #[op]
+      let some _ ← synthInstance? cop | failure
+      pure true
+    catch _ =>
+      pure false
+  modify fun _ =>
+    { ccs with
+      canOps := ccs.canOps.insert op op
+      opInfo := ccs.opInfo.insert op b }
+  return bif b then some op else none
+
+open MessageData in
+/-- Given `lhs`, `rhs`, and `header := "my header:"`, Trace `my header: lhs = rhs`. -/
+def dbgTraceACEq (header : String) (lhs rhs : ACApps) : CCM Unit := do
+  let ccs ← get
+  trace[Debug.Meta.Tactic.cc.ac]
+    group (ofFormat (header ++ .line) ++ ccs.ppACApps lhs ++
+      ofFormat (.line ++ "=" ++ .line) ++ ccs.ppACApps rhs)
+
+open MessageData in
+/-- Trace the state of AC module. -/
+def dbgTraceACState : CCM Unit := do
+  let ccs ← get
+  trace[Debug.Meta.Tactic.cc.ac] group ("state: " ++ nest 6 ccs.ppAC)
+
+/-- Return the proof of `e₁ = e₂` using `ac_rfl` tactic. -/
+def mkACProof (e₁ e₂ : Expr) : MetaM Expr := do
+  let eq ← mkEq e₁ e₂
+  let .mvar m ← mkFreshExprSyntheticOpaqueMVar eq | failure
+  AC.rewriteUnnormalized m
+  let pr ← instantiateMVars (.mvar m)
+  mkExpectedTypeHint pr eq
+
+/-- Given `tr := t*r` `sr := s*r` `tEqs : t = s`, return a proof for `tr = sr`
+
+    We use `a*b` to denote an AC application. That is, `(a*b)*(c*a)` is the term `a*a*b*c`. -/
+def mkACSimpProof (tr t s r sr : ACApps) (tEqs : DelayedExpr) : MetaM DelayedExpr := do
+  if tr == t then
+    return tEqs
+  else if tr == sr then
+    let some tre := tr.toExpr | failure
+    DelayedExpr.ofExpr <$> mkEqRefl tre
+  else
+    let .apps op _ := tr | failure
+    let some re := r.toExpr | failure
+    let opr := op.app re
+    let some te := t.toExpr | failure
+    let rt := opr.app te
+    let some se := s.toExpr | failure
+    let rs := mkApp2 op re se
+    let rtEqrs := DelayedExpr.congrArg opr tEqs
+    let some tre := tr.toExpr | failure
+    let trEqrt ← mkACProof tre rt
+    let some sre := sr.toExpr | failure
+    let rsEqsr ← mkACProof rs sre
+    return .eqTrans (.eqTrans trEqrt rtEqrs) rsEqsr
```

Comment:
Let us reorder the statements for symmetry.
```suggestion
    let .apps op _ := tr | failure
    let some re := r.toExpr | failure
    let some te := t.toExpr | failure
    let some se := s.toExpr | failure
    let some tre := tr.toExpr | failure
    let some sre := sr.toExpr | failure
    let opr := op.app re -- `(*) r`
    let rt := opr.app te -- `r * t`
    let rs := mkApp2 op re se -- `r * s`
    let rtEqrs := DelayedExpr.congrArg opr tEqs
    let trEqrt ← mkACProof tre rt
    let rsEqsr ← mkACProof rs sre
    return .eqTrans (.eqTrans trEqrt rtEqrs) rsEqsr
```
Perhaps additionally symmetrize the construction of `rt` and `rs`:
```suggestion
    let .apps op _ := tr | failure
    let some re := r.toExpr | failure
    let some te := t.toExpr | failure
    let some se := s.toExpr | failure
    let some tre := tr.toExpr | failure
    let some sre := sr.toExpr | failure
    let opr := op.app re -- `(*) r`
    let rt := mkApp2 op re te -- `r * t`
    let rs := mkApp2 op re se -- `r * s`
    let rtEqrs := DelayedExpr.congrArg opr tEqs
    let trEqrt ← mkACProof tre rt
    let rsEqsr ← mkACProof rs sre
    return .eqTrans (.eqTrans trEqrt rtEqrs) rsEqsr
```